src/components/chatbox/chatbox.tsx

```diff
@@ -0,0 +1,550 @@
+	const mutation = useMutation<Response, Response, MutationParams>({
+		onMutate: async () => {
+			// clear previous response
+			setStreamedText('...')
+			setErrorText('')
+			setMessageId('')
+		},
+		onError: async (error) => {
+			console.log('onError', error)
+			switch (error.status) {
+				case 400:
+				case 401:
+				case 403: {
+					setErrorText(`${error.status} ${error.statusText}`)
+					break
+				}
+				case 429: {
+					const resetAtSec = mutation.error.headers.get('x-ratelimit-reset')
+					const resetAtMs = Number(resetAtSec) * 1000
+					const diffMs = resetAtMs - Date.now()
+
+					const errorMessage = `Too many requests. Please try again in ${ms(
+						diffMs,
+						{
+							long: true,
+						}
+					)}.`
+					setErrorText(errorMessage)
+					break
+				}
+				default: {
+					setErrorText(`${error.status} ${error.statusText}`)
+					break
+				}
+			}
+		},
+		mutationFn: async ({ value: task, conversationId, parentMessageId }) => {
+			// call our edge function which is capable of streaming
+			const request = async (jwt: string) => {
+				const res = await fetch('/api/chat/route', {
+					method: 'POST',
+					headers: {
+						'Content-Type': 'application/json',
+						Authorization: `Bearer ${token}`,
+					},
+					body: JSON.stringify({ task, conversationId, parentMessageId }),
+				})
+				return res
+			}
+
+			let response = await request(token)
+
+			if (response.ok) {
+				// messageId and conversationId are noops until we are ready for multi-message conversations
+				setConversationId(response.headers.get('x-conversation-id'))
+				setMessageId(response.headers.get('x-message-id'))
+				return response
+			} else {
+				// if the response is not ok
+				// if the response is a 401, we need to refresh the access token and retry once.
+				// otherwise, throw the response
+				if (response.status == 401) {
+					const { accessToken } = await update()
+					response = await request(accessToken)
+					if (response.ok) {
+						return response
+					}
+				}
+				throw response
+			}
+		},
+	})
+
+	// A stream reader
+	const [reader, setReader] =
+		useState<ReadableStreamDefaultReader<Uint8Array> | null>(null)
+	// Function to stop the stream
+	const stopStream = () => {
+		reader?.cancel()
+		mutation.reset()
+	}
+
+	// when the mutation / POST is successful
+	useEffect(() => {
+		if (mutation.data?.ok) {
+			const stream = mutation.data.body
+			const _reader = stream.getReader()
+			setReader(_reader)
+
+			// transform stream to async iterable;
+			// each chunk is transformed into a SSE object
+			const iter = streamToAsyncIterable(_reader)
+
+			// Self-invoking async function to read the stream
+			;(async () => {
+				// Reset streamed text
+				setStreamedText('')
+				setIsReading(true)
+
+				// This is our contract with the backend
+				const messageFormat = z.object({
+					content: z.string(),
+				})
+				const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms))
```

Comment:
can you move this out of the useEffect since it doesn't need to capture any values? also can you rename `r` to `resolve`? Took me a second to realize that's what it was.

Reply:
done in 2827e4706